<h1> Effect of Quick Mindfulness on Cognitive Functioning </h1>

<p>
    In this experiment, we collected participants' sustained attention and working memory scores during the initial session. Subsequently, we divided them into two groups: an experimental group and a control group. The experimental group engaged in a 10-minute quick mindfulness exercise, while the control group did not receive any treatment but instead waited for 10 minutes until the next session occurred. During the subsequent session, we once again assessed the sustained attention and working memory scores of all participants to determine if the quick mindfulness exercise have any effects.

</p>

<b> Hypotheses </b>

1. Sustained Attention scores of the experimental group will enhance in later session
2. Working Memory scores of the experimental group will enhance in later session

<b> Data Analysis </b>

1. We intend to use Repeated Measures ANOVA for Sustained Attention and Working Memory. If assumptions are confirmed.
2. If assumptions are violated then sequence of Wilcoxon signed-rank test will be conducted.

In [1]:
import pandas as pd
import numpy as np
import statsmodels.api as sm
from statsmodels.formula.api import ols
from statsmodels.stats.anova import AnovaRM
import pingouin as pg
import scipy.stats as st
import gspread

> Data inserting </br>
> SA stands for Sustained Attention</br>
> WM stands for Working Memory
> Time stands for the phase (pretest & posttest)
> Group stands for the group (control & experimental)

In [6]:
sh = gspread.service_account(filename="C:/Users/DELL/Desktop/Data Analysis Projects/iot-jupyter-lab-114b492fbed8.json")
sheet = sh.open("Q_P")
work_sheet = sheet.worksheet("SART_2_back_task")
df = pd.read_csv('SART_2_back_task.csv')

<h2> Normality </h2>

> Here we used K-S test for testing the normality of our data

In [7]:
norm_l = list({
'norm_c_sa_pretest' : st.kstest(df['SA'][[df['Time'] == 'pretest'] and df['group'] == 'control'], 'norm'),
               
'norm_c_sa_posttest' : st.kstest(df['SA'][[df['Time'] == 'posttest'] and df['group'] == 'control'], 'norm'),
               
'norm_e_sa_pretest' : st.kstest(df['SA'][[df['Time'] == 'pretest'] and df['group'] == 'experimental'], 'norm'),
               
'norm_e_sa_posttest' : st.kstest(df['SA'][[df['Time'] == 'posttest'] and df['group'] == 'experimental'], 'norm'),
    
'norm_c_wm_pretest' : st.kstest(df['WM'][[df['Time'] == 'pretest'] and df['group'] == 'control'], 'norm'),
               
'norm_c_wm_posttest' : st.kstest(df['WM'][[df['Time'] == 'posttest'] and df['group'] == 'control'], 'norm'),
               
'norm_e_wm_pretest' : st.kstest(df['WM'][[df['Time'] == 'pretest'] and df['group'] == 'experimental'], 'norm'),
               
'norm_e_wm_posttest' : st.kstest(df['WM'][[df['Time'] == 'posttest'] and df['group'] == 'experimental'], 'norm')
              }.items())


for i in range(len(norm_l)) :
    print ( norm_l[i][0], " -  k : ", norm_l[i][1][0], ', p-value : ', norm_l[i][1][1], '\n')

norm_c_sa_pretest  -  k :  0.5844186937516695 , p-value :  6.6866756876742305e-74 

norm_c_sa_posttest  -  k :  0.5844186937516695 , p-value :  6.6866756876742305e-74 

norm_e_sa_pretest  -  k :  0.6236191730674514 , p-value :  2.170631954683758e-85 

norm_e_sa_posttest  -  k :  0.6236191730674514 , p-value :  2.170631954683758e-85 

norm_c_wm_pretest  -  k :  0.5810072106386759 , p-value :  5.903291651692851e-73 

norm_c_wm_posttest  -  k :  0.5810072106386759 , p-value :  5.903291651692851e-73 

norm_e_wm_pretest  -  k :  0.6236191730674514 , p-value :  2.170631954683758e-85 

norm_e_wm_posttest  -  k :  0.6236191730674514 , p-value :  2.170631954683758e-85 



<b> Normality of the data isn't confirmed </b>
> As our data is in proportion, arcsine transformation can be helpful to acheive normality

In [8]:
df['arcSA'] = np.arcsin(df['SA'])
df['arcWM'] = np.arcsin(df['WM'])

In [9]:
norm_l = list({
'norm_c_sa_pretest' : st.kstest(df['SA'][[df['Time'] == 'pretest'] and df['group'] == 'control'], 'norm'),
               
'norm_c_sa_posttest' : st.kstest(df['SA'][[df['Time'] == 'posttest'] and df['group'] == 'control'], 'norm'),
               
'norm_e_sa_pretest' : st.kstest(df['SA'][[df['Time'] == 'pretest'] and df['group'] == 'experimental'], 'norm'),
               
'norm_e_sa_posttest' : st.kstest(df['SA'][[df['Time'] == 'posttest'] and df['group'] == 'experimental'], 'norm'),
    
'norm_c_wm_pretest' : st.kstest(df['WM'][[df['Time'] == 'pretest'] and df['group'] == 'control'], 'norm'),
               
'norm_c_wm_posttest' : st.kstest(df['WM'][[df['Time'] == 'posttest'] and df['group'] == 'control'], 'norm'),
               
'norm_e_wm_pretest' : st.kstest(df['WM'][[df['Time'] == 'pretest'] and df['group'] == 'experimental'], 'norm'),
               
'norm_e_wm_posttest' : st.kstest(df['WM'][[df['Time'] == 'posttest'] and df['group'] == 'experimental'], 'norm')
              }.items())


for i in range(len(norm_l)) :
    print ( norm_l[i][0], " -  k : ", norm_l[i][1][0], ', p-value : ', norm_l[i][1][1], '\n')

norm_c_sa_pretest  -  k :  0.5844186937516695 , p-value :  6.6866756876742305e-74 

norm_c_sa_posttest  -  k :  0.5844186937516695 , p-value :  6.6866756876742305e-74 

norm_e_sa_pretest  -  k :  0.6236191730674514 , p-value :  2.170631954683758e-85 

norm_e_sa_posttest  -  k :  0.6236191730674514 , p-value :  2.170631954683758e-85 

norm_c_wm_pretest  -  k :  0.5810072106386759 , p-value :  5.903291651692851e-73 

norm_c_wm_posttest  -  k :  0.5810072106386759 , p-value :  5.903291651692851e-73 

norm_e_wm_pretest  -  k :  0.6236191730674514 , p-value :  2.170631954683758e-85 

norm_e_wm_posttest  -  k :  0.6236191730674514 , p-value :  2.170631954683758e-85 



<h3> Arcsine transformed data didn't confirmed normality either </h3>

<h2> Non Paracmetric Approach </h2>
    As our data violated the assumptions. we intend to conduct a series of wilcoxon test to identify the diffrences of mean

In [10]:
# Perform the Wilcoxon signed-rank test

def wilcoxres(data1,data2):
    statistic, p_value = st.wilcoxon(data1, data2)

    # Output the results
    print(f"Wilcoxon Statistic: {statistic}")
    print(f"P-value: {p_value}")

    # Interpret the results
    alpha = 0.05  # Set your desired significance level
    if p_value < alpha:
        return "Reject the null hypothesis. There is a significant difference between the samples."
        
    else:
        return "Fail to reject the null hypothesis. There is no significant difference between the samples."

In [11]:
#Sustained Attention in Control Group
wilcoxres(

df['SA'][(df['Time']=='pretest') & (df['group'] =='control')],
df['SA'][(df['Time']=='posttest') & (df['group'] =='control')]
)

Wilcoxon Statistic: 1867.0
P-value: 0.00016238563142671853


'Reject the null hypothesis. There is a significant difference between the samples.'

In [12]:
#Sustained Attention in Experimental Group
wilcoxres(

df['SA'][(df['Time']=='pretest') & (df['group'] =='experimental')],
df['SA'][(df['Time']=='posttest') & (df['group'] =='experimental')]
)

Wilcoxon Statistic: 0.0
P-value: 2.5625676925228196e-20


'Reject the null hypothesis. There is a significant difference between the samples.'

In [13]:
#Working Memory in Control Group
wilcoxres(

df['WM'][(df['Time']=='pretest') & (df['group'] =='control')],
df['WM'][(df['Time']=='posttest') & (df['group'] =='control')]
)

Wilcoxon Statistic: 2946.5
P-value: 0.7519026284212349


'Fail to reject the null hypothesis. There is no significant difference between the samples.'

In [14]:
#Working Memory in Experimental Group
wilcoxres(

df['WM'][(df['Time']=='pretest') & (df['group'] =='experimental')],
df['WM'][(df['Time']=='posttest') & (df['group'] =='experimental')]
)

Wilcoxon Statistic: 17.0
P-value: 6.459597227398733e-20


'Reject the null hypothesis. There is a significant difference between the samples.'

<h2> Results </h2>

# Analysis of Wilcoxon Rank Sum Test

## Sustained Attention

- In the Control group, there are significant differences between the pretest and posttest scores for Sustained Attention:
  - **Wilcoxon Statistic:** 1867.0
  - **P-value:** 0.0001624

- In the Experimental group, significant differences between the pretest and posttest scores for Sustained Attention were observed:
  - **Wilcoxon Statistic:** 0.0
  - **P-value:** 2.5626e-20

## Working Memory

- In the Control group, there is no significant difference between the pretest and posttest scores for Working Memory:
  - **Wilcoxon Statistic:** 2946.5
  - **P-value:** 0.7519

- In the Experimental group, significant differences between the pretest and posttest scores for Working Memory were found:
  - **Wilcoxon Statistic:** 17.0
  - **P-value:** 6.4596e-20


In [15]:
df.groupby(['group','Time'])[['SA','WM']].mean()

SA        WM
group        Time                        
control      posttest  0.499027  0.477788
             pretest   0.489381  0.481018
experimental posttest  0.745752  0.733628
             pretest   0.498274  0.486991

Here, we examined the performance of two groups, the Control and Experimental groups, on measures of *Sustained Attention (SA) and Working Memory (WM)* at two different time points, pretest and posttest. For the Control group, the mean SA scores were approximately *0.489* at pretest and *0.499* at posttest, indicating a slight improvement. In contrast, the Experimental group demonstrated more substantial progress, with mean SA scores of approximately  *0.498* at pretest and a notable increase to about *0.746* at posttest. The Wilcoxon Test results for Sustained Attention in the Control group indicated significant differences between the pretest and posttest scores. However, the descriptive statistics underscore the more significant improvement observed in the Experimental group, providing strong support for our hypothesis that quick mindfulness interventions enhance Sustained Attention.

Turning to Working Memory, the Control group displayed mean scores of roughly *0.481* at pretest and *0.478* at posttest, suggesting minimal change. Meanwhile, the Experimental group exhibited substantial improvement, with mean scores of about *0.487* at pretest and a considerable increase to around *0.734* at posttest. Both the descriptive statistics and the Wilcoxon Test results for Working Memory in the Experimental group align in showing a significant enhancement in scores following the quick mindfulness intervention.